File: cypress_automation/cypress/pages/loginPage.js
class LoginPage {

    visit() {
        cy.visit("https://www.saucedemo.com/");
    }

    enterUsername(username) {
        cy.get("#user-name").clear().type(username);
    }

    enterPassword(password) {
        cy.get("#password").clear().type(password);
    }

    clickLogin() {
        cy.get("#login-button").click();
    }

    login(username, password) {
        this.visit();
        this.enterUsername(username);
        this.enterPassword(password);
        this.clickLogin();
    }

    verifyLoginSuccess() {
        cy.url().should("include", "/inventory.html");
    }

    verifyLoginFailure() {
        cy.get('[data-test="error"]').should("be.visible");
    }

    logout() {
        cy.get("#react-burger-menu-btn").click();
        cy.get("#logout_sidebar_link").click();
    }
}

export default LoginPage;

File: selenium_automation/pages/login_page.py
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC


class LoginPage:

    USERNAME = (By.ID, "user-name")
    PASSWORD = (By.ID, "password")
    LOGIN_BUTTON = (By.ID, "login-button")
    ERROR_MESSAGE = (By.CSS_SELECTOR, "h3[data-test='error']")

    def __init__(self, driver):
        self.driver = driver
        self.wait = WebDriverWait(driver, 20)

    def enter_username(self, username):
        username_field = self.wait.until(
            EC.visibility_of_element_located(self.USERNAME)
        )
        username_field.clear()
        username_field.send_keys(username)

    def enter_password(self, password):
        password_field = self.wait.until(
            EC.visibility_of_element_located(self.PASSWORD)
        )
        password_field.clear()
        password_field.send_keys(password)

    def click_login(self):
        self.wait.until(
            EC.element_to_be_clickable(self.LOGIN_BUTTON)
        ).click()

    def login(self, username, password):
        self.enter_username(username)
        self.enter_password(password)
        self.click_login()

        # Wait until inventory page is fully loaded
        self.wait.until(
            EC.visibility_of_element_located(
                (By.ID, "inventory_container")
            )
        )

    def get_error_message(self):
        return self.wait.until(
            EC.visibility_of_element_located(self.ERROR_MESSAGE)
        ).text